onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -128,8 +128,8 @@ Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level,
       // implementations for all ONNX ops. If the node/op is from a different op domain or if the CPU implementation
       // does not support the specific input type(s) required by the node (currently we only support a subset of
       // types in some CPU kernels) then we can't proceed with constant folding for the node.
-      auto ep_type = node->GetExecutionProviderType();
-      bool cpu_ep = ep_type == kCpuExecutionProvider;
+      auto& ep_type = node->GetExecutionProviderType();
+      bool cpu_ep = ep_type.empty() || ep_type == kCpuExecutionProvider;
```

Comment:
Found a bug in detecting whether a node is assigned to the CPU ep. This bug prevents non-onnx domain ops with ep_type == "" from being constant folded (e.g., com.microsoft.QuantizeLinear).

Reply:
I thought that `ep_type` being empty meant that the node is not assigned to an EP yet.

perhaps this early check `if (!cpu_ep && node->Domain() != kOnnxDomain)` is too conservative. after all, there's a CPU kernel lookup that happens later and that will actually indicate whether there is a CPU kernel available. wonder if we can just skip the domain check?

note that I'm not that familiar with this code and may very well be missing something.

Reply:
AFAIK empty == unassigned. 

This is definitely confusing. Constant folding typically runs in level 1 which is prior to any node assignments, and the EP type would be empty. At one point we did run level 1 optimizers again after partitioning, but we don't currently unless we do an NHWC transformation. i.e. I'm not sure when cpu_ep would be 'true' for a typical model with no NHWC transformation.

Default post-partitioning optimization:

https://github.com/microsoft/onnxruntime/blob/cbaa00839177650073da298d7693e7e42f6940e1/onnxruntime/core/session/inference_session.cc#L1051-L1056

Exception to re-run level 1 optimization when we do layout transform:

https://github.com/microsoft/onnxruntime/blob/cbaa00839177650073da298d7693e7e42f6940e1/onnxruntime/core/session/inference_session.cc#L997-L1010

As a side note that possibly means we're losing some potential constant folding if the model requires a CUDA kernel to do the constant folding. e.g. if fp16 data is involved the CPU EP kernel might not support that but the CUDA kernel would.

Reply:
Yeah, I was wrong about the empty `ep_type`.

As I look at this more and more, it really does look like this check is unnecessary since we're also checking if the CPU kernel exists. I refactored and pushed a commit to see if CIs are ok. 

Perhaps we can discuss this offline.

Reply:
Do we still need to check the node domain? What if an operator only has cuda or rocm implementation?

Reply:
There is an explicit check for the existence of a CPU kernel here: https://github.com/microsoft/onnxruntime/pull/17127/files#diff-8184f402ae2f6aec257554809f988eb357153603d5ccaf77597c77b0119baf83R208

 If a CPU kernel implementation does not exist (regardless of domain), then the loop moves on to the next candidate node.